File: web/src/game/world.js
/**
 * World clock and weather.
 *
 * Time and weather are not decoration: they are two of the four multipliers in
 * the catch table, and they drive the sky, water colour and wave energy. One
 * in-game day is compressed to a few real minutes so a session actually crosses
 * a dawn.
 */

import { EV } from '../core/events.js';
import { TIME_PHASES, WEATHER, WEATHER_BY_ID, phaseForHour } from '../data/spots.js';
import { clamp01, damp, lerp } from '../core/loop.js';

/** Real seconds per in-game hour. */
export const HOUR_SECONDS = 26;

export class World {
  constructor(rng, bus, spot) {
    this.rng = rng;
    this.bus = bus;
    this.spot = spot;
    // Start mid-morning: dawn is the prettiest light in the game but it is also
    // the dimmest, and a first-run player should not open into a dark screen.
    this.hour = 8.6;
    this.day = 1;
    this.phase = phaseForHour(this.hour);
    this.weather = WEATHER_BY_ID.clear;
    this.nextWeatherIn = rng.float(90, 220);
    this.weatherBlend = 1;     // 0..1 transition progress into `weather`
    this.prevWeather = this.weather;

    // Continuous drivers the renderer reads.
    this.wind = spot.windBase;
    this.chop = 0.5;
    this.light = 1;
    this.rain = 0;
    this.sunAngle = 0;
  }

  setSpot(spot) {
    this.spot = spot;
    this.wind = spot.windBase;
  }

  /** Bite activity multiplier: fish feed hardest on a falling light level. */
  activity() {
    const p = this.phase;
    const base = p.id === 'dawn' || p.id === 'dusk' ? 1.35 : p.id === 'night' ? 0.85 : 1.0;
    const w = this.weather.id === 'rain' ? 1.25 : this.weather.id === 'storm' ? 0.7 : 1.0;
    return base * w;
  }

  update(dt) {
    // --- clock -----------------------------------------------------------
    const prevPhase = this.phase.id;
    this.hour += dt / HOUR_SECONDS;
    if (this.hour >= 24) { this.hour -= 24; this.day++; }
    this.phase = phaseForHour(this.hour);
    if (this.phase.id !== prevPhase) {
      this.bus.emit(EV.TIME_PHASE, { phase: this.phase, hour: this.hour, day: this.day });
    }

    // --- weather ---------------------------------------------------------
    this.nextWeatherIn -= dt;
    if (this.nextWeatherIn <= 0) {
      const next = this.rng.weighted(WEATHER.map((w) => [w.id, w.chance]));
      if (next && next !== this.weather.id) {
        this.prevWeather = this.weather;
        this.weather = WEATHER_BY_ID[next];
        this.weatherBlend = 0;
        this.bus.emit(EV.WEATHER_CHANGE, { weather: this.weather, from: this.prevWeather });
      }
      this.nextWeatherIn = this.rng.float(110, 260);
    }
    this.weatherBlend = Math.min(1, this.weatherBlend + dt / 12);

    // --- continuous drivers ----------------------------------------------
    const w = this.weather, pw = this.prevWeather, b = this.weatherBlend;
    const windTarget = this.spot.windBase * lerp(pw.wind, w.wind, b);
    const chopTarget = lerp(pw.chop, w.chop, b);
    const rainTarget = lerp(pw.rain, w.rain, b);

    // Sun elevation: a smooth arc that peaks at noon and goes negative at night.
    const t = (this.hour - 6) / 12;                     // 0 at 06:00, 1 at 18:00
    this.sunAngle = Math.sin(t * Math.PI);
    const daylight = clamp01(this.sunAngle * 1.35 + 0.06);
    const lightTarget = daylight * lerp(pw.light, w.light, b);

    this.wind = damp(this.wind, windTarget, 0.6, dt);
    this.chop = damp(this.chop, chopTarget, 0.5, dt);
    this.rain = damp(this.rain, rainTarget, 0.8, dt);
    this.light = damp(this.light, lightTarget, 1.2, dt);
  }

  /** Surface noise the bite system uses: chop and rain both mask a clumsy cast. */
  surfaceNoise() {
    return clamp01(this.chop * 0.22 + this.rain * 0.30 + this.wind * 0.10);
  }

  clockString() {
    const h = Math.floor(this.hour);
    const m = Math.floor((this.hour - h) * 60);
    return `${String(h).padStart(2, '0')}:${String(m).padStart(2, '0')}`;
  }

  telemetry() {
    return {
      hour: this.hour, day: this.day, clock: this.clockString(),
      phase: this.phase, weather: this.weather,
      wind: this.wind, chop: this.chop, light: this.light, rain: this.rain,
      activity: this.activity(),
    };
  }
}

export { TIME_PHASES, WEATHER };

File: unity/Pancing/Assets/Pancing/Sim/Game/World.cs
using System;
using System.Collections.Generic;

namespace Pancing.Sim
{
    /// <summary>
    /// World clock and weather — a port of web/src/game/world.js.
    ///
    /// Time and weather are not decoration: they are two of the four multipliers
    /// in the catch table, and they drive the sky, water colour and wave energy.
    /// One in-game day is compressed to a few real minutes so a session actually
    /// crosses a dawn.
    /// </summary>
    public sealed class World
    {
        /// <summary>Real seconds per in-game hour.</summary>
        public const double HourSeconds = 26;

        private readonly Rng _rng;
        private readonly EventBus _bus;
        private readonly SpotDb _db;
        private readonly List<double> _weatherWeights = new List<double>();

        public Spot Spot;
        public double Hour = 8.6;
        public int Day = 1;
        public TimePhase Phase;
        public WeatherSpec Weather;
        public WeatherSpec PrevWeather;
        public double NextWeatherIn;
        /// <summary>0..1 transition progress into `Weather`.</summary>
        public double WeatherBlend = 1;

        // Continuous drivers the renderer reads.
        public double Wind;
        public double Chop = 0.5;
        public double Light = 1;
        public double Rain;
        public double SunAngle;

        public World(Rng rng, EventBus bus, SpotDb db, Spot spot)
        {
            _rng = rng;
            _bus = bus;
            _db = db;
            Spot = spot;

            // Start mid-morning: dawn is the prettiest light in the game but it is
            // also the dimmest, and a first-run player should not open into a dark
            // screen.
            Hour = 8.6;
            Day = 1;
            Phase = db.PhaseForHour(Hour);
            Weather = db.GetWeather("clear");
            PrevWeather = Weather;
            NextWeatherIn = _rng.Float(90, 220);
            WeatherBlend = 1;

            Wind = spot.WindBase;

            foreach (var w in db.Weathers) _weatherWeights.Add(w.Chance);
        }

        public void SetSpot(Spot spot)
        {
            Spot = spot;
            Wind = spot.WindBase;
        }

        /// <summary>Bite activity multiplier: fish feed hardest on a falling light level.</summary>
        public double Activity()
        {
            string p = Phase?.Id;
            double bas = (p == "dawn" || p == "dusk") ? 1.35 : (p == "night" ? 0.85 : 1.0);
            double w = Weather?.Id == "rain" ? 1.25 : (Weather?.Id == "storm" ? 0.7 : 1.0);
            return bas * w;
        }

        public void Update(double dt)
        {
            // --- clock ---------------------------------------------------------
            string prevPhase = Phase?.Id;
            Hour += dt / HourSeconds;
            if (Hour >= 24) { Hour -= 24; Day++; }
            Phase = _db.PhaseForHour(Hour);
            if (Phase?.Id != prevPhase)
            {
                _bus?.Emit(EV.TimePhaseChanged, new PhaseChange { Phase = Phase, Hour = Hour, Day = Day });
            }

            // --- weather -------------------------------------------------------
            NextWeatherIn -= dt;
            if (NextWeatherIn <= 0)
            {
                int idx = _rng.WeightedIndex(_weatherWeights);
                if (idx >= 0)
                {
                    var next = _db.Weathers[idx];
                    if (next.Id != Weather.Id)
                    {
                        PrevWeather = Weather;
                        Weather = next;
                        WeatherBlend = 0;
                        _bus?.Emit(EV.WeatherChange, new WeatherChange { Weather = Weather, From = PrevWeather });
                    }
                }
                NextWeatherIn = _rng.Float(110, 260);
            }
            WeatherBlend = Math.Min(1, WeatherBlend + dt / 12);

            // --- continuous drivers --------------------------------------------
            var w2 = Weather; var pw = PrevWeather; double b = WeatherBlend;
            double windTarget = Spot.WindBase * MathUtil.Lerp(pw.Wind, w2.Wind, b);
            double chopTarget = MathUtil.Lerp(pw.Chop, w2.Chop, b);
            double rainTarget = MathUtil.Lerp(pw.Rain, w2.Rain, b);

            // Sun elevation: a smooth arc that peaks at noon and goes negative at
            // night.
            double t = (Hour - 6) / 12;                     // 0 at 06:00, 1 at 18:00
            SunAngle = Math.Sin(t * Math.PI);
            double daylight = MathUtil.Clamp01(SunAngle * 1.35 + 0.06);
            double lightTarget = daylight * MathUtil.Lerp(pw.Light, w2.Light, b);

            Wind = MathUtil.Damp(Wind, windTarget, 0.6, dt);
            Chop = MathUtil.Damp(Chop, chopTarget, 0.5, dt);
            Rain = MathUtil.Damp(Rain, rainTarget, 0.8, dt);
            Light = MathUtil.Damp(Light, lightTarget, 1.2, dt);
        }

        /// <summary>Surface noise the bite system uses: chop and rain both mask a clumsy cast.</summary>
        public double SurfaceNoise() => MathUtil.Clamp01(Chop * 0.22 + Rain * 0.30 + Wind * 0.10);

        public string ClockString()
        {
            int h = (int)Math.Floor(Hour);
            int m = (int)Math.Floor((Hour - h) * 60);
            return $"{h:00}:{m:00}";
        }

        public struct PhaseChange { public TimePhase Phase; public double Hour; public int Day; }
        public struct WeatherChange { public WeatherSpec Weather; public WeatherSpec From; }

        public struct Telemetry
        {
            public double Hour;
            public int Day;
            public string Clock;
            public TimePhase Phase;
            public WeatherSpec Weather;
            public double Wind, Chop, Light, Rain, Activity;
        }

        public Telemetry GetTelemetry() => new Telemetry
        {
            Hour = Hour, Day = Day, Clock = ClockString(),
            Phase = Phase, Weather = Weather,
            Wind = Wind, Chop = Chop, Light = Light, Rain = Rain,
            Activity = Activity(),
        };
    }
}
